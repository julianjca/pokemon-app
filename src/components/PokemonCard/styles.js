import styled from "@emotion/styled";

export const Card = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),
    0 4px 6px -2px rgba(0, 0, 0, 0.05);
  height: 180px;
  border-radius: 1.5rem;
`;
