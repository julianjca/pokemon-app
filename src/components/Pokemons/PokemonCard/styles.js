import styled from "@emotion/styled";

export const Card = styled.a`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),
    0 4px 6px -2px rgba(0, 0, 0, 0.05);
  border-radius: 1.5rem;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  transition: 0.1 all ease-in-out;
  background: #ffffff;
  color: ${(props) => props.theme.colors.primary};
  margin: 0 10px 10px 0;
  height: 160px;
  width: 160px;

  & > span {
    font-size: 0.9rem;
    margin-top: 2px;
  }

  @media (min-width: ${(props) => props.theme.breakpoints.md}) {
    height: 180px;
    width: 180px;
    margin: 0 15px 15px 0;
  }

  &:hover {
    transform: translateY(-3px);
    transition: 0.1s all ease-in-out;
  }
`;

export const Pokeball = styled.img`
  position: absolute;
  opacity: 0.08;
  bottom: -58px;
  right: -60px;
  height: 200px;
`;

export const PokemonName = styled.div`
  font-family: ${(props) => props.theme.fonts.primary};
  font-weight: bold;
  text-transform: capitalize;
`;
