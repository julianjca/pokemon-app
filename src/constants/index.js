export const ADD_POKEMON = "ADD_POKEMON";
export const RELEASE_POKEMON = "RELEASE_POKEMON";
export const SET_POKEMONS = "SET_POKEMONS";
