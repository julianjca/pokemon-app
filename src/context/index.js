import React, { useEffect, createContext, useReducer } from "react";

const PokemonStateContext = createContext();
const PokemonDispatchContext = createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case "ADD_POKEMON": {
      const pokemons = [...state.pokemons].concat(action.pokemon);
      localStorage.setItem("pokemons", JSON.stringify(pokemons));
      return { pokemons };
    }
    case "REMOVE_POKEMON": {
      return { pokemons: state.count - 1 };
    }
    case "SET_POKEMONS": {
      return { pokemons: action.pokemons };
    }
    default:
      console.log("error");
  }
};

const PokemonStateProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, {
    pokemons: [],
  });

  useEffect(() => {
    if (typeof window !== undefined) {
      const pokemons = localStorage.getItem("pokemons");
      if (pokemons) {
        dispatch({
          type: "SET_POKEMONS",
          pokemons: JSON.parse(pokemons),
        });
      }
    }
  }, []);

  return (
    <PokemonStateContext.Provider value={state}>
      <PokemonDispatchContext.Provider value={dispatch}>
        {children}
      </PokemonDispatchContext.Provider>
    </PokemonStateContext.Provider>
  );
};

const usePokemonState = () => {
  const context = React.useContext(PokemonStateContext);

  return context;
};

const usePokemonDispatch = () => {
  const context = React.useContext(PokemonDispatchContext);

  return context;
};

export { PokemonStateProvider, usePokemonState, usePokemonDispatch };
