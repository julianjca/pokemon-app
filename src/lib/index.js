export const makeSentenceCase = (str) => {
  return str
    .split(" ")
    .map((w) => w[0].toUpperCase() + w.substr(1).toLowerCase())
    .join(" ");
};

export const countPokemon = (pokemons, name) => {
  return pokemons.reduce(
    (acc, value) => ({
      ...acc,
      [value.pokemonName]: (acc[value.pokemonName] || 0) + 1,
    }),
    {}
  )[name];
};

export const uniqBy = (arr, key) => {
  const seen = {};

  return arr.filter((item) => {
    const k = key(item);
    return seen.hasOwnProperty(k) ? false : (seen[k] = true);
  });
};

export const getCatchingOdds = () => {
  return getCatchingOdds;
};
