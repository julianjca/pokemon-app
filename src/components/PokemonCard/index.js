import React from "react";
import Link from "next/link";

import { Card, Pokeball } from "./styles";

const PokemonCard = ({ name, image }) => {
  return (
    <Link href={`/pokemons/${name}`}>
      <Card>
        <img src={image} alt={name} />
        <div>{name}</div>
        <Pokeball src="/pokeball.svg" />
      </Card>
    </Link>
  );
};

export default PokemonCard;
