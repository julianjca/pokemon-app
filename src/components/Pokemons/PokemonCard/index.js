import React from "react";
import Link from "next/link";

import { Card, Pokeball, PokemonName } from "./styles";

const PokemonCard = ({ name, image }) => {
  return (
    <Link href={`/pokemons/${name}`}>
      <Card>
        <img src={image} alt={name} />
        <PokemonName>{name}</PokemonName>
        <Pokeball src="/pokeball.svg" />
      </Card>
    </Link>
  );
};

export default PokemonCard;
