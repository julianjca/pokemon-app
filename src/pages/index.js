import React from "react";
import Head from "next/head";

import Pokemons from "../components/Pokemons";

export default function Home() {
  return (
    <>
      <Head>
        <title>Pokemon App</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <Pokemons />
    </>
  );
}
