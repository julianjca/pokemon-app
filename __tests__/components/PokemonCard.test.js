/* eslint-disable testing-library/prefer-screen-queries */
import React from "react";
import { matchers } from "@emotion/jest";

import { render } from "../../test-utils";
import PokemonCard from "../../src/components/Pokemons/PokemonCard";
import * as lib from "../../src/lib";

expect.extend(matchers);

HTMLCanvasElement.prototype.getContext = () => {
  // return whatever getContext has to return
};

jest.mock("../../src/lib");

describe("PokemonCard", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });
  test("renders the component", () => {
    const element = render(<PokemonCard name="bulbasaur" />);

    expect(element).toMatchSnapshot();
  });

  test("renders pokemon data", () => {
    const { getByText } = render(<PokemonCard name="bulbasaur" />);

    expect(getByText("bulbasaur")).toBeInTheDocument();
  });

  test("renders pokemon count", () => {
    const countPokemon = jest.spyOn(lib, "countPokemon").mockReturnValueOnce(2);

    const { getByText } = render(<PokemonCard name="bulbasaur" />);

    expect(getByText("You have: 2")).toBeInTheDocument();

    countPokemon.mockRestore();
  });
});
