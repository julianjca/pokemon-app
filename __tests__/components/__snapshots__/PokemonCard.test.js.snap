// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  border-radius: 1.5rem;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  -webkit-transition: 0.1 all ease-in-out;
  transition: 0.1 all ease-in-out;
  background: #ffffff;
  color: #193569;
  margin: 0 10px 10px 0;
  height: 160px;
  width: 160px;
}

.emotion-0>span {
  font-size: 0.9rem;
  margin-top: 2px;
}

@media (min-width: 720px) {
  .emotion-0 {
    height: 180px;
    width: 180px;
    margin: 0 15px 15px 0;
  }
}

.emotion-0:hover {
  -webkit-transform: translateY(-3px);
  -moz-transform: translateY(-3px);
  -ms-transform: translateY(-3px);
  transform: translateY(-3px);
  -webkit-transition: 0.1s all ease-in-out;
  transition: 0.1s all ease-in-out;
}

.emotion-2 {
  font-family: Inter,sans-serif;
  font-weight: bold;
  text-transform: capitalize;
}

.emotion-4 {
  position: absolute;
  opacity: 0.08;
  bottom: -58px;
  right: -60px;
  height: 200px;
}

<body>
    <div>
      <a
        class="emotion-0 emotion-1"
      >
        <img
          alt="bulbasaur"
        />
        <div
          class="emotion-2 emotion-3"
        >
          bulbasaur
        </div>
        <span>
          You have: 
          0
        </span>
        <img
          class="emotion-4 emotion-5"
          src="/pokeball.svg"
        />
      </a>
    </div>
  </body>,
  "container": .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  border-radius: 1.5rem;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  -webkit-transition: 0.1 all ease-in-out;
  transition: 0.1 all ease-in-out;
  background: #ffffff;
  color: #193569;
  margin: 0 10px 10px 0;
  height: 160px;
  width: 160px;
}

.emotion-0>span {
  font-size: 0.9rem;
  margin-top: 2px;
}

@media (min-width: 720px) {
  .emotion-0 {
    height: 180px;
    width: 180px;
    margin: 0 15px 15px 0;
  }
}

.emotion-0:hover {
  -webkit-transform: translateY(-3px);
  -moz-transform: translateY(-3px);
  -ms-transform: translateY(-3px);
  transform: translateY(-3px);
  -webkit-transition: 0.1s all ease-in-out;
  transition: 0.1s all ease-in-out;
}

.emotion-2 {
  font-family: Inter,sans-serif;
  font-weight: bold;
  text-transform: capitalize;
}

.emotion-4 {
  position: absolute;
  opacity: 0.08;
  bottom: -58px;
  right: -60px;
  height: 200px;
}

<div>
    <a
      class="emotion-0 emotion-1"
    >
      <img
        alt="bulbasaur"
      />
      <div
        class="emotion-2 emotion-3"
      >
        bulbasaur
      </div>
      <span>
        You have: 
        0
      </span>
      <img
        class="emotion-4 emotion-5"
        src="/pokeball.svg"
      />
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
